### PR TITLE
fix: add toggle eye in password bb-145

### PR DESCRIPTION
**Pull request by yulyaolshanska**
## Summary
Added toggle eye icon for password in Sign Up / Sign In forms
Fixed bug 145, 146

### Sign Up / Sign In

image

image

### CODE REVIEW

File: apps/frontend/src/pages/auth/components/sign-in-form/sign-in-form.tsx [hunk 1]
```diff
@@ -20,6 +20,7 @@ const SignInForm: React.FC<Properties> = ({ onSubmit }: Properties) => {
 		defaultValues: DEFAULT_SIGN_IN_PAYLOAD,
 		validationSchema: userSignInValidationSchema,
 	});
+	const [isPasswordVisible, setIsPasswordVisible] = useState(false);
```
Comment (fshabanov) [inline apps/frontend/src/pages/auth/components/sign-in-form/sign-in-form.tsx:23]:
```suggestion
	const [isPasswordVisible, setIsPasswordVisible] = useState<boolean>(false);
```

File: apps/frontend/src/pages/auth/components/sign-up-form/sign-up-form.tsx [hunk 2]
```diff
@@ -19,6 +19,7 @@ const SignUpForm: React.FC<Properties> = ({ onSubmit }: Properties) => {
 		defaultValues: DEFAULT_SIGN_UP_PAYLOAD,
 		validationSchema: userSignUpValidationSchema,
 	});
+	const [isPasswordVisible, setIsPasswordVisible] = useState(false);
```
Comment (fshabanov) [inline apps/frontend/src/pages/auth/components/sign-up-form/sign-up-form.tsx:22]:
```suggestion
	const [isPasswordVisible, setIsPasswordVisible] = useState<boolean>(false);
```

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/libs/enums/icon-name.enum.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+const IconName = {
```
Comment (fshabanov) [inline apps/frontend/src/libs/enums/icon-name.enum.ts:1]: We already have type IconName, we don't need enum

File: apps/frontend/src/libs/components/input/input.tsx [hunk 4]
```diff
@@ -68,12 +74,25 @@ const Input = <T extends FieldValues>({
...
+							type={type === "password" && isDisplayedValue ? "text" : type}
+						/>
+						{type === "password" && onToggle && (
+							<button
+								className={styles["toggle-password-icon"]}
+								onClick={onToggle}
+								type="button"
+							>
+								<Icon
+									name={isDisplayedValue ? IconName.CROSSED_EYE : IconName.EYE}
```
Comment (fshabanov) [inline apps/frontend/src/libs/components/input/input.tsx:91]: iconName should be sent to Input component as a prop. Other inputs can also have icons

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/libs/enums/icon-name.enum.ts (hunk 3 above)
Comment (yulyaolshanska) [reply apps/frontend/src/libs/enums/icon-name.enum.ts:1]: so I can just pass the name as a string, not a constant? 
```
iconName={isPasswordVisible ? "crossedEye" : "eye"}
```
Or is it better to leave the enum and change the type to this?
```
import { ICON_NAMES } from "~/libs/enums/enums.js";

type IconName = typeof ICON_NAMES[keyof typeof ICON_NAMES];
```

File: apps/frontend/src/libs/components/input/input.tsx [hunk 5]
```diff
@@ -22,8 +27,11 @@ type Properties<T extends FieldValues> = {
 const Input = <T extends FieldValues>({
 	control,
 	errors,
+	iconName,
+	isDisplayedValue = false,
```
Comment (WilhelmWesser) [inline apps/frontend/src/libs/components/input/input.tsx:31]: Let's move ```isDisplayedValue``` right before placeholder parameter as function parameters with default values are usually moved to the end of parameters

**Review by WilhelmWesser — CHANGES_REQUESTED** [review]

File: apps/frontend/src/libs/components/input/input.tsx (hunk 5 above)
Comment (fshabanov) [inline apps/frontend/src/libs/components/input/input.tsx:31]: Handle this outside - send a needed type based on this value

File: apps/frontend/src/libs/components/input/input.tsx [hunk 6]
```diff
@@ -22,8 +27,11 @@ type Properties<T extends FieldValues> = {
...
 	label,
 	name,
+	onToggle,
```
Comment (fshabanov) [inline apps/frontend/src/libs/components/input/input.tsx:34]:
```suggestion
	onIconClick,
```
It will not always be toggle

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/libs/components/input/input.tsx [hunk 7]
```diff
@@ -68,12 +76,23 @@ const Input = <T extends FieldValues>({
...
-						{...field}
-						placeholder={placeholder}
-						type={type}
-					/>
+					<>
+						<input
+							className={getValidClassNames(styles["input-field"])}
+							{...field}
+							placeholder={placeholder}
+							type={type === "password" && isDisplayedValue ? "text" : type}
```
Comment (revenets) [inline apps/frontend/src/libs/components/input/input.tsx:84]: Please, move it to a separate variable

**Review by revenets — CHANGES_REQUESTED** [review]

**Review by revenets — APPROVED** [review]

File: apps/frontend/src/libs/components/input/input.tsx [hunk 8]
```diff
@@ -33,6 +39,7 @@ const Input = <T extends FieldValues>({
 	const error = errors?.[name]?.message;
 	const hasError = Boolean(error);
 	const isRadioWithOptions = type === "radio" && options?.length;
+	const shouldShowIcon = iconName && typeof onIconClick === "function";
```
Comment (fshabanov) [inline apps/frontend/src/libs/components/input/input.tsx:42]:
```suggestion
	const shouldShowIcon = Boolean(iconName);
```
Icon doesn't have to be clickable

File: apps/frontend/src/libs/components/input/input.tsx [hunk 9]
```diff
@@ -68,12 +75,23 @@ const Input = <T extends FieldValues>({
...
+							{...field}
+							placeholder={placeholder}
+							type={type}
+						/>
+						{shouldShowIcon && (
+							<button
+								className={styles["toggle-password-icon"]}
+								onClick={onIconClick}
+								type="button"
+							>
```
Comment (fshabanov) [inline apps/frontend/src/libs/components/input/input.tsx:90]: Use existing Button component

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/libs/components/input/input.tsx (hunk 8 above)
Comment (yulyaolshanska) [reply apps/frontend/src/libs/components/input/input.tsx:42]: But if I remove typeof onIconClick === "function", I get error 
  Types of property 'onClick' are incompatible.
  Type '(() => void) | undefined' is not assignable to type 'MouseEventHandler<HTMLButtonElement>'.

Comment (fshabanov) [reply apps/frontend/src/libs/components/input/input.tsx]: Create another function within Icon component:
```
const handleIconClick = (): void => {
  onIconClick?.()
}
```
And use it

Comment (yulyaolshanska) [reply apps/frontend/src/libs/components/input/input.tsx]:
```
	onClick?: (() => void) | undefined;

``` 
I know that onClick? means that onClick can be undefined. But without " | undefined " I get a type error.

Comment (fshabanov) [reply apps/frontend/src/libs/components/input/input.tsx]: It's expected to be this way

File: apps/frontend/src/libs/components/button/button.tsx [hunk 10]
```diff
@@ -3,16 +3,20 @@ import { getValidClassNames } from "~/libs/helpers/helpers.js";
 import styles from "./styles.module.css";
 
 type Properties = {
+	children?: React.ReactNode;
 	isFluid?: boolean;
+	isIconButton?: boolean;
 	isPrimary?: boolean;
-	label: string;
+	label?: string;
```
Comment (fshabanov) [inline apps/frontend/src/libs/components/button/button.tsx:10]:
```suggestion
	label: string;
```
Label is mandatory. Add `hasVisuallyHiddenLabel` prop and if it is true, use `visually-hidden` class to hide label

File: apps/frontend/src/libs/components/button/button.tsx [hunk 11]
```diff
@@ -3,16 +3,20 @@ import { getValidClassNames } from "~/libs/helpers/helpers.js";
 import styles from "./styles.module.css";
 
 type Properties = {
+	children?: React.ReactNode;
 	isFluid?: boolean;
+	isIconButton?: boolean;
```
Comment (fshabanov) [inline apps/frontend/src/libs/components/button/button.tsx:8]: I don't think we need that. If we have icon, then it is icon button. Add `iconName` prop here and avoid `children`

File: apps/frontend/src/libs/components/input/input.tsx [hunk 12]
```diff
@@ -33,6 +39,7 @@ const Input = <T extends FieldValues>({
 	const error = errors?.[name]?.message;
 	const hasError = Boolean(error);
 	const isRadioWithOptions = type === "radio" && options?.length;
+	const shouldShowIconButton = iconName && typeof onIconClick === "function";
```
Comment (fshabanov) [inline apps/frontend/src/libs/components/input/input.tsx:42]:
```suggestion
	const shouldShowIconButton = iconName;
```
Icon can be non-clickable

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/libs/components/input/input.tsx (hunk 8 above)
Comment (fshabanov) [reply apps/frontend/src/libs/components/input/input.tsx]: Yes, we should have `| undefined` there

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/libs/components/button/button.tsx [hunk 13]
```diff
@@ -1,33 +1,48 @@
...
+	hasVisuallyHiddenLabel?: boolean;
+	iconName?: IconName;
 	isFluid?: boolean;
 	isPrimary?: boolean;
 	label: string;
-	onClick?: React.MouseEventHandler<HTMLButtonElement>;
+	onClick?: (() => void) | undefined;
 	type?: "button" | "submit";
-	variant: "dark" | "secondary";
+	variant?: "dark" | "icon" | "secondary";
```
[the remark this answers is not in the document]
Comment (yulyaolshanska) [reply apps/frontend/src/libs/components/button/button.tsx:15]: I saw variant="dark" many times, but we dont have styles for this variant

[the remark this answers is not in the document]
Comment (fshabanov) [reply apps/frontend/src/libs/components/button/button.tsx:15]: If no styles for it, please, remove it

File: apps/frontend/src/libs/components/input/input.tsx [hunk 14]
```diff
@@ -68,12 +74,24 @@ const Input = <T extends FieldValues>({
...
+							className={getValidClassNames(styles["input-field"])}
+							{...field}
+							placeholder={placeholder}
+							type={type}
+						/>
+						{iconName && (
+							<Button
+								hasVisuallyHiddenLabel
+								iconName={iconName}
+								label=""
```
Comment (fshabanov) [inline apps/frontend/src/libs/components/input/input.tsx:88]: Add a label here. It should be added for accessibility purposes

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/libs/components/input/input.tsx (hunk 14 above)
Comment (yulyaolshanska) [reply apps/frontend/src/libs/components/input/input.tsx:88]: Can i add label like label="Icon" ?

**Review by fshabanov — APPROVED** [review]

**Pull request merged**